design/20190708.certificate-request-crd.md

```diff
@@ -174,6 +188,190 @@ type ObjectReference struct {
+conditions of the same name. The purpose of these conditions is so that users,
+or "approvers", are able to mark a request as either Approved or Denied. This
+condition gates a signer from signing the request. A signer waits for a
+CertificateRequest to have an Approved condition before signing. A
+CertificateRequest with a Denied condition will never be signed.
+
+##### Behaviour
+
+The Approved and Denied conditions are two distinct condition types on the
+CertificateRequest. These conditions must only have the status of True, and
+are mutually exclusive (i.e. a CertificateRequest cannot have an Approved and
+Denied condition simultaneously). This behaviour is enforced in the cert-manager
+validating admission webhook.
+
+An "approver" is an entity that is responsible for setting the Approved/Denied
+conditions. It is up to the approver's implementation as to what
+CertificateRequests are managed by that approver.
+
+The Reason field of the Approved/Denied condition should be set to *who* set the
+condition. Who can be interpreted however makes sense to the approver
+implementation. For example, it may include the API group of an approving policy
+controller, or the client agent of a manual request.
+
+The Message field of the Approved/Denied condition should be set to *why* the
+condition is set. Again, why can be interpreted however makes sense to the
+implementation of the approver. For example, the name of the resource that
+approves this request, the violations which caused the request to be denied, or
+the team to who manually approved the request.
+
+When a CertificateRequest has been Denied, it is the responsibility of the
+referenced issuer to then add a Ready condition with the status of "False",
+along with a relevant Reason and Message.
+
+
+##### RBAC
+
+Approved and Denied conditions are set by requesting against the `/status`
+endpoint of the CertificateRequest resource. This is a divergence of the
+certificates.k8s.io CertificateSigningRequest resource that utilises `approval`,
+as CustomResourceDefinitions may [only define `/status` or `/scale`
+sub-resources](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#subresources).
+Approvers must therefore have permissions to update the status sub-resource of
+CertificateRequests. Approvers missing this permissions will have their request
+rejected by the API server.
+
+Setting the Approved or Denied conditions are restricted by the approver having
+sufficient RBAC permissions. These permissions are based upon the request
+itself- specifically the request's IssuerRef:
+
+```yaml
+apiGroups: ["cert-manager.io"]
+resources: ["signers"]
+verbs: ["approve"]
+resourceNames:
+ # namesapced signers
+ - "<signer-resource-name>.<signer-group>/<signer-namespace>.<signer-name>"
+ # cluster scoped signers
+ - "<signer-resource-name>.<signer-group>/<signer-name>"
+ # all signers of this resource name
+ - "<signer-resource-name>.<signer-group>/*"
+```
```

Comment:
Might be worth changing "resource name" to "resource type" since it seems to be the official term for it (cf. [api-concepts](https://kubernetes.io/docs/reference/using-api/api-concepts/))

I also propose to clarify what each `resourceName` syntax does:

~~~suggestion
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
rules:
- apiGroups: ["cert-manager.io"]
  resources: ["signers"]
  verbs: ["approve"]
  resourceNames:
   # To allow a namespaced signer, use the following syntax:
   - "<signer-resource-type>.<signer-group>/<signer-namespace>.<signer-resource-name>"
   # To allow a cluster-scoped signer, use the following syntax:
   - "<signer-resource-type>.<signer-group>/<signer-resource-name>"
   # To allow a all the signers for a given group:
   - "<signer-resource-type>.<signer-group>/*"
```
~~~